sklearn/tree/_classes.py

```diff
@@ -905,6 +852,12 @@ class DecisionTreeClassifier(ClassifierMixin, BaseDecisionTree):
             0.93...,  0.93...,  1.     ,  0.93...,  1.      ])
     """
 
+    _parameter_constraints = {
+        **BaseDecisionTree._parameter_constraints,
+        "criterion": [StrOptions({"gini", "entropy", "log_loss"})],
```

Comment:
Looks like an instance of `Criterion` is also valid, but in a kind of non documented way, maybe for internal purpose (seems not even tested). Should we expose it in the error message or instead make it possible to hide some valid options from the error message like it's proposed for the string options in https://github.com/scikit-learn/scikit-learn/pull/23459 ? @thomasjpfan @ogrisel

Reply:
Historically, I think `Criterion` instances were unofficially supported for users that want to provided their own Criterion objects with the same interface (from Cython).

If we want to maintain backward compatibility with this unofficially supported feature, then we need to allow for `Criterion` subclasses, i.e. check `isinstance(self.criterion, Criterion)`. Can the validation framework allow for such a constraint?

Reply:
I also think that we need to keep backward compat. For the constraint it's easy, we just need to add `Criterion` as a possibiliy: `"criterion": [StrOptions({"gini", "entropy", "log_loss"}), Criterion],`.
My concern was that it then would appear in the error message `criterion must be a str among {"gini", "entropy", "log_loss"} or a Criterion instance` which makes it kind of official and I'm not sure we want to make it official. So do you think it's fine to expose it in the error message ? or do you think it would be better to hide it like we did for the "deprecated" string ?

Reply:
Ah I think it's better to hide it. Which means there needs to be another "internal" mechanism for types? 😅

Reply:
Yes :smile:
I have an idea that should make it simple. I'll submit a PR soon

Reply:
@thomasjpfan I opened https://github.com/scikit-learn/scikit-learn/pull/23558 to make it possible to mark a constraint as internal.